File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])

        visit = set()
        ans = 0

        def dfs(i, j):
            if i not in range(rows) or j not in range(cols):
                return 0
            
            if grid[i][j] == 0:
                return 0
            
            if (i, j) in visit:
                return 0

            visit.add((i, j))
            return 1 + dfs(i + 1, j) + dfs(i - 1, j) + dfs(i, j + 1) + dfs(i, j - 1)
        
        for row in range(rows):
            for col in range(cols):
                count = dfs(row, col)
                ans = max(ans, count)
        
        return ans

File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-0.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<bool> rows(m);
        vector<bool> cols(n);

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (matrix[r][c] == 0) {
                    rows[r] = true;
                    cols[c] = true;
                }
            }
        }

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (rows[r] || cols[c]) {
                    matrix[r][c] = 0;
                }
            }
        }
    }
};